### PR TITLE
geph: init at 4.5.0-beta.1

### PR DESCRIPTION
**Pull request by jslll135**
###### Description of changes
geph: init at 4.5.0-beta.1
A modular Internet censorship circumvention system designed specifically to deal with national filtering.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jslll135** [comment]
@NickCao Could you please review this PR? (Sorry for pinging you, just thought that maybe a v2ray package maintainer is interested in this.)

**Comment by NickCao** [comment]
A few nitpicking: the commit message should be in the format: `<pname>: init at <version>`, and cleaning up the history into three commits: add maintainer, add the node package, add the rust packages would make it look leaner.

File: pkgs/applications/networking/geph/cargo.patch [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+index 27e6f3c..bf26c97 100644
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -108,3 +108,6 @@ opt-level=3
+ debug=0
+ lto=true
+ 
++[patch.crates-io.conv]
++git = 'https://github.com/DanielKeep/rust-conv.git'
++rev = '9b7e41a5e1c0508ed58cf41066744d706bba10ed'
```
Comment (NickCao) [inline pkgs/applications/networking/geph/cargo.patch:25]: And why is this patch needed, adding random patches harms maintainability in the long term.

Comment (jslll135) [reply pkgs/applications/networking/geph/cargo.patch:25]: I don't really know. There is a dep chain: geph4 -> geph4-aioutils -> dnsclient -> dnssector -> chomp -> conv. Both chomp and conv are not maintained for years. chomp requires conv 0.3.3 with std feature however conv 0.3.3 does not has that feature. I don't know why it builds fine with cargo build but can't resolve the dep with nix-build. So I have to patch it.

Comment (NickCao) [reply pkgs/applications/networking/geph/cargo.patch:25]: Maybe cargo build recreates the lockfile with correct dependency versions? In that case I would suspect this as a upstream issue worth reporting.

**Review by NickCao — APPROVED** [review]

File: pkgs/applications/networking/geph/cargo.patch [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+\ No newline at end of file
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/geph/cargo.patch:26]: New lines are important for git.

File: pkgs/applications/networking/geph/default.nix [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      rev = "v${version}";
+      hash = "sha256-FtdXl+29VVJxmKvtc03PFji/ilkJ8XTymUqjPfBh5hk=";
+    };
+
+    cargoPatches = [ ./cargo.patch ];
+
+    cargoHash = "sha256-wJE9QW9X/LDfa32pUK7oUc+w1MSd1fVodk73ZcfZhSI=";
+
+    meta = geph-meta // {
+      mainProgram = pname;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/geph/default.nix:41]: That is the default.

File: pkgs/applications/networking/geph/default.nix [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      inherit src;
+      sourceRoot = "source/${pname}";
+
+      cargoHash = "sha256-NutGGc+xXJ8CDkBXSlE0r5O19q/epmoNSUvrOXmlNKA=";
+    };
+
+    gephgui-wry = rustPlatform.buildRustPackage rec {
+      pname = "gephgui-wry";
+      version = "2c9f681a003e78583f4846abe9a6085098717a9f";
+
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/geph/default.nix:71]: Please just the unstable format.

Comment (jslll135) [reply pkgs/applications/networking/geph/default.nix:70]: Done.

**Comment by alienzj** [comment]
Can we merge it?

**Pull request closed without merging**